### PR TITLE
✨ Implement ForEachReducer

### LINKED ISSUE #53: Add ForEach composition method
## Description
The Composable Architecture has this handy [ForEach](https://github.com/pointfreeco/swift-composable-architecture/blob/main/Sources/ComposableArchitecture/Reducer/Reducers/ForEachReducer.swift) method of composing reducers. See this in action in [Todos sample app](https://github.com/pointfreeco/swift-composable-architecture/blob/20089ee985b04b1ae82e9742aa9d9c8f044700c5/Examples/Todos/Todos/Todos.swift#L97). 
We should be able to do the same, although it will probably look a bit uglier.

## Definition of done
- [ ] ForEach reducer extension method
- [ ] ForEachReducer
- [ ] Tests are written

### PR DESCRIPTION
**Pull request by semanticer**
## Summary
This PR is introducing `ForEachReducer` with `forEach`, `forEachMap` and `forEachList` extension methods. TCA's rough equivalent is [here](https://github.com/pointfreeco/swift-composable-architecture/blob/main/Sources/ComposableArchitecture/Reducer/Reducers/ForEachReducer.swift).

## Relationships

Closes #53

## Tests
✅  I could technically extract stuff from `ForEachMapTests`  and  `ForEachListTests` and create some common ancestor to reduce code repetition. I'm just not sure if I can do it without making it very hard to read.

### CODE REVIEW

File: komposable-architecture/src/main/java/com/toggl/komposable/extensions/ReducerExtensions.kt [hunk 1]
```diff
@@ -62,3 +63,87 @@ fun <LocalState, GlobalState, LocalAction, GlobalAction>
...
+            }
+            mapToParentState(state, newElementList)
+        },
+    )
+
+/**
+ * Embeds a child reducer within a parent domain, allowing it to operate on elements of a collection
+ * within the parent's state.
+ *
+ * For example, if a parent feature manages an array of child states, you can utilize the `forEach`
```
Comment (heytherewill) [inline komposable-architecture/src/main/java/com/toggl/komposable/extensions/ReducerExtensions.kt:127]: NIT

```suggestion
 * For example, if a parent feature manages an array of child states, you can utilize the [forEach]
 
```

File: komposable-architecture/src/test/java/com/toggl/komposable/reducer/forEach/ForEachTestUtils.kt [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+sealed class ParentAction {
+    data class EditText(val text: String) : ParentAction()
+    data class ElementActionWrapper(val id: Int, val elementAction: ElementAction) : ParentAction()
+}
+
+sealed class ElementAction {
+    data class EditText(val text: String) : ElementAction()
+}
+
+data class ElementState(val elementText: String)
```
Comment (heytherewill) [inline komposable-architecture/src/test/java/com/toggl/komposable/reducer/forEach/ForEachTestUtils.kt:16]: Calling this class `ElementState` makes it really hard to wrap one's head around what should happen. Using a less generic name (like just `Element` or maybe even `Item`) could improve readability for the tests.

Even better would be to change these into a concrete example, so real world usage would be easier to grasp for someone trying to understand this concept from the tests. TimeEntries feel like a very natural example. Another good one would be a list of songs/albums

**Review by heytherewill — APPROVED** [review]
I love the concept of this, but the syntax is so convoluted that it loses some of its appeal (the elegance of embedding the child reducer). We should definitely merge this, but thinking about how to improve it with KSP should be a task for 1.0

Approving for good measure, though I think my non-NIT comment should be addressed

File: komposable-architecture/src/test/java/com/toggl/komposable/reducer/forEach/ForEachTestUtils.kt (hunk 2 above)
Comment (semanticer) [reply komposable-architecture/src/test/java/com/toggl/komposable/reducer/forEach/ForEachTestUtils.kt:16]: `ElementState` is derived from the generic name used in forEach. But yeah, tests might not be the best way to wrap one's head around this concept; that's where sample apps and case studies come in handy. I've started working on the new TODO app, where this should be demonstrated quite effectively.

**Pull request merged**